File: MapboxVelocity/MapboxVelocity/Core/Design/DesignSystem.swift
//
//  DesignSystem.swift
//  MapboxVelocity
//
//  Created by Shaikat on 4/3/26.
//

import UIKit
import MapboxMaps

// MARK: - Design System Protocol (Interface Segregation Principle)

// Consumers depend on this protocol, not the concrete enum.

protocol DesignSystemProviding {
    var primaryColor: UIColor { get }
    var secondaryColor: UIColor { get }
    var surfaceColor: UIColor { get }
    var mapStyleURI: StyleURI { get }
}

// MARK: - Concrete implementation from UI/UX Design
struct VelocityDesignSystem: DesignSystemProviding {
    // Backgrounds
    let backgroundColor  = UIColor(hex: "#121212")
    let surfaceColor     = UIColor(hex: "#1C1C1E")
    let surface2Color    = UIColor(hex: "#2C2C2E")
    
    // Brand
    let primaryColor     = UIColor(hex: "#007AFF")  // Main CTA
    let secondaryColor   = UIColor(hex: "#34C759")  // Success/confirm
    let tertiaryColor    = UIColor(hex: "#FF9500")  // Warning/zone 3
    
    // Text
    let textPrimary      = UIColor.white
    let textSecondary    = UIColor(hex: "#8E8E93")
    let textTertiary     = UIColor(hex: "#636366")
    
    // Map style matching dark theme
    let mapStyleURI: StyleURI = .dark
    
    // Spacing scale (4pt base grid)
    enum Spacing {
        static let xs: CGFloat  = 4
        static let sm: CGFloat  = 8
        static let md: CGFloat  = 16
        static let lg: CGFloat  = 24
        static let xl: CGFloat  = 32
        static let xxl: CGFloat = 48
    }
    
    // Corner radius tokens
    enum Radius {
        static let sm: CGFloat  = 8
        static let md: CGFloat  = 12
        static let lg: CGFloat  = 16
        static let xl: CGFloat  = 24
    }
    
    // Typography
    enum Typography {
        static let headline = UIFont.systemFont(ofSize: 28, weight: .bold)
        static let title    = UIFont.systemFont(ofSize: 20, weight: .semibold)
        static let body     = UIFont.systemFont(ofSize: 16, weight: .regular)
        static let caption  = UIFont.systemFont(ofSize: 12, weight: .regular)
        static let button   = UIFont.systemFont(ofSize: 16, weight: .semibold)
    }
}

File: MapboxVelocityAndroid/app/src/main/java/com/mapboxvelocity/Core/Design/DesignSystem.kt
package com.mapboxvelocity.Core.Design

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

interface DesignSystemProviding {
    val primaryColor: Color
    val secondaryColor: Color
    val tertiaryColor: Color
    val backgroundColor: Color
    val surfaceColor: Color
    val surface2Color: Color
    val textPrimary: Color
    val textSecondary: Color
    val textTertiary: Color
}

object VelocityDesignSystem : DesignSystemProviding {
    override val primaryColor    = Color(0xFF007AFF)
    override val secondaryColor  = Color(0xFF34C759)
    override val tertiaryColor   = Color(0xFFFF9500)
    override val backgroundColor = Color(0xFF121212)
    override val surfaceColor    = Color(0xFF1C1C1E)
    override val surface2Color   = Color(0xFF2C2C2E)
    override val textPrimary     = Color(0xFFFFFFFF)
    override val textSecondary   = Color(0xFF8E8E93)
    override val textTertiary    = Color(0xFF636366)

    // Card surface used in map search & bottom sheet
    val cardColor = Color(0xFF1a1f2e)

    object Spacing {
        val xs: Dp = 4.dp
        val sm: Dp = 8.dp
        val md: Dp = 16.dp
        val lg: Dp = 24.dp
        val xl: Dp = 32.dp
        val xxl: Dp = 48.dp
    }

    object Radius {
        val sm: Dp = 8.dp
        val md: Dp = 12.dp
        val lg: Dp = 16.dp
        val xl: Dp = 24.dp
    }

    object Typography {
        val headline = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title    = TextStyle(fontSize = 20.sp, fontWeight = FontWeight.SemiBold)
        val body     = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Normal)
        val caption  = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Normal)
        val button   = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
    }
}
